### PR TITLE
Add audio focus support in Android platform.

### PR DESCRIPTION
**Pull request by shenguojun**
To avoid every music app playing at the same time, add focus support in Android platform.

### CODE REVIEW

**Comment by HugoHeneault** [comment]
Am I right or it also supports `ACTION_AUDIO_BECOMING_NOISY` so when we unplug an headset the sound can stop? :)

**Comment by shenguojun** [comment]
> Am I right or it also supports `ACTION_AUDIO_BECOMING_NOISY` so when we unplug an headset the sound can stop? :)

I will mute the sond when unplug the headset, but I think it will not continue when you plug againt.

File: android/src/main/java/xyz/luan/audioplayers/WrappedMediaPlayer.java [hunk 1]
```diff
@@ -356,4 +400,51 @@ private void setAttributes(MediaPlayer player, Context context) {
         }
     }
 
-}
+    /**
+     * 音频焦点改变的音量变化控制
```
Comment (luanpotter) [inline android/src/main/java/xyz/luan/audioplayers/WrappedMediaPlayer.java:404]: what are these squares??

File: android/src/main/java/xyz/luan/audioplayers/WrappedSoundPool.java [hunk 2]
```diff
@@ -328,4 +376,52 @@ private UnsupportedOperationException unsupportedOperation(String message) {
         return new UnsupportedOperationException("LOW_LATENCY mode does not support: " + message);
     }
 
-}
+    /**
+     * 音频焦点改变的音量变化控制
```
Comment (luanpotter) [inline android/src/main/java/xyz/luan/audioplayers/WrappedSoundPool.java:380]: ditto

**Comment by ryanheise** [comment]
One of the problems I see is that if an app uses multiple audio plugins within the same app, and each one of those audio plugins takes on the policy decisions for audio focus / audio mixing responsibility upon itself, you may end up with different plugins overwriting each other's global settings. And even if all audio plugins agree to make this configurable, it is still not desirable for different audio plugins to blindly overwrite these global settings previously written by each individual plugin, even if they are overwritten with equal values set independently by all other audio plugins. It is redundant and a duplication of responsibilities.

I have run into this problem before where one audio plugin set the AVAudioSession category on iOS and then another audio plugin in the same app set its own AVAudioSession category overwriting the choice of the first plugin. In fact, what was appropriate for my app was a mix of categories and options that were specific to my app, and the scope of this policy decision logically ought to have been at the app level, not at the individual audio plugin level. After all, each audio plugin in an app is effectively a "guest" of the shared audio session which is a singleton instance shared by the app. Assumptions made by one plugin will therefore interfere with assumptions made by other audio plugins in the same app.

On Android, the equivalent API is AudioManager, and the same sort of problem can arise.

You may want to consider looking into [audio_session](https://pub.dev/packages/audio_session) which allows configuring these policy decisions at the app level rather than the individual audio plugin level.

**Comment by erickzanardo** [comment]
This PR is quite outdated and haven't got updates in a long time. Closing this. If the author wants to continue working on this, just lets us know and we can re open the PR.

**Pull request closed without merging**